tests/mypy/outputs/plugin-fail.txt

```diff
@@ -25,4 +26,7 @@
 197: note: Possible overload variant:
 197: note:     def dataclass(*, init: bool = ..., repr: bool = ..., eq: bool = ..., order: bool = ..., unsafe_hash: bool = ..., frozen: bool = ..., config: Optional[Type[Any]] = ...) -> Callable[[Type[Any]], Type[Dataclass]]
 197: note:     <1 more non-matching overload not shown>
-219: error: Property "y" defined in "FrozenModel" is read-only  [misc]
\ No newline at end of file
+219: error: Property "y" defined in "FrozenModel" is read-only  [misc]
+239: error: Incompatible types in assignment (expression has type "str", variable has type "int")  [assignment]
```

Comment:
Looking at implementation, this message does make sense. However it may not be obvious, especially when you're using some custom functions (you don't see a return type just by glancing at the line). Can we make it more obvious by changing the message? Would it be possible to fail fast at [pydantic.mypy:160](https://github.com/samuelcolvin/pydantic/pull/3430/files#diff-469037bbe55bbf9aa359480a16040d368c676adad736e133fb07e5e20d6ac523R160) with some custom message like 
```py
f"Incompatible type in `default_factory` ({default_factory_type.fullname} has return type {default_factory_type.ret_type!r}, variable has type {...})"
```

Reply:
Thank you and totally agree, will follow your suggestion using the `overlapping_type` API from mypy

Reply:
I tried to implement your suggestion, but I find a hard roadblock due to the separation of parsing expression into types and the function/class plugin hook.

What I currently do is "fake" the return value of the `Field` function in the mypy function hook, so that the `mypy` plugin will autonomously raise errors for type mismatch.

What you suggest cannot be done in the function hook: I do not have any information there on the type of the variable I am assigning the function to, therefore I cannot compare the types and raise the error.

Similarly, I tried to move this check to the class hook (where I have information about the variable types); the main problem there is that I do not have info on the function `Field` and its argument types, since the type checking comes later during the function hook :( .

I agree that it would be better to be more specific, but nonetheless I think that it is an improvement to add this check for the `default_factory`, and the message is not unclear 🤔 

What do you think? In case, I can also revert the change on the `default_factory` and keep only the check for the `default` argument.

Reply:
Your current implementation is definitely a big improvement. If it's not trivial to customize the message, as you say, I'd stick with current method for now.